### PR TITLE
disallow entities in list type: fixes #1435, fixes #1758

### LINKED ISSUE #1435: list type should not allow for entities in the list
The following model should fail but doesn't
```
entity ListContainer:
	list lst
end

implement ListContainer using std::none

x = ListContainer()
x.lst = [x]
```

### LINKED ISSUE #1758: infinite recursion in TypedList.type_string()?
https://github.com/inmanta/inmanta/blob/592faa8f98334b77de17cf05945d0cc79952880c/src/inmanta/ast/type.py#L356

I believe this line is meant to call `self.basetype.type_string()`

### PR DESCRIPTION
**Pull request by sanderr**

### CODE REVIEW

**Comment by sanderr** [comment]
There's part of this I'm not really proud of, but given the current type system this is the best I could come up with. If you've got suggestions for a better approach, please share them.

**Comment by sanderr** [comment]
TODO: use `TypedList`s for `multi` types

File: src/inmanta/ast/statements/define.py [hunk 1]
```diff
@@ -148,14 +163,20 @@ def evaluate(self) -> None:
...
             add_attributes: Dict[str, Attribute] = {}
             for attribute in self.attributes:
-                attr_type = self.namespace.get_type(attribute.type)
+                attr_type: Type = attribute.type.get_type(self.namespace)
                 if not isinstance(attr_type, (Type, type)):
                     raise TypingException(self, "Attributes can only be a type. Entities need to be defined as relations.")
 
                 name = str(attribute.name)
-                attr_obj = Attribute(entity_type, attr_type, name, attribute.multi, attribute.nullable)
+                attr_obj = Attribute(
```
Comment (wouterdb) [inline src/inmanta/ast/statements/define.py]: perhaps attribute can be similarly be cleaned up by passing a Type?

File: src/inmanta/ast/type.py [hunk 2]
```diff
@@ -390,46 +239,61 @@ def validate(cls, value):
...
+    def normalize(self) -> None:
+        self.element_type.normalize()
+
+    def validate(self, value: Optional[object]) -> bool:
+        if not List.validate(self, value):
+            return False
+
+        assert isinstance(value, list)
+        for element in value:
+            self.element_type.validate(element)
```
Comment (wouterdb) [inline src/inmanta/ast/type.py]: no return?

File: src/inmanta/ast/type.py [hunk 3]
```diff
@@ -390,46 +239,61 @@ def validate(cls, value):
...
+    """
+        This class represents a list type in the configuration model. (instances represent instances)
+    """
 
-    @classmethod
-    def validate(cls, value):
+    def __init__(self) -> None:
+        Type.__init__(self)
+
+    def validate(self, value: Optional[object]) -> bool:
```
Comment (wouterdb) [inline src/inmanta/ast/type.py]: The existing 

```
if value is None:
            return True	       
```

is a bit scary, but it is existing, so we can let irt be for now.

File: src/inmanta/ast/type.py [hunk 4]
```diff
@@ -445,18 +309,76 @@ def validate(cls, value):
...
+        self.types: PythonList[Type] = types
+
+    def validate(self, value: object) -> bool:
+        for typ in self.types:
+            try:
+                if typ.validate(value):
+                    return True
+            except RuntimeException:
+                pass
+        raise RuntimeException(None, "Invalid value '%s', expected Literal" % value)
```
Comment (wouterdb) [inline src/inmanta/ast/type.py]: Why Literal?

File: src/inmanta/ast/statements/define.py (hunk 1 above)
Comment (sanderr) [reply src/inmanta/ast/statements/define.py]: I agree, but the type is also used for ResultVariable creation so it wouldn't be a straightforward replacement. I'll make a seperate ticket for this.

Comment (sanderr) [reply src/inmanta/ast/statements/define.py]: see #1770

**Pull request merged**